deps/uvwasi/src/fd_table.c

```diff
@@ -522,3 +325,67 @@ uvwasi_errno_t uvwasi_fd_table_remove(uvwasi_t* uvwasi,
   uv_rwlock_wrunlock(&table->rwlock);
   return err;
 }
+
+
+uvwasi_errno_t uvwasi_fd_table_renumber(struct uvwasi_s* uvwasi,
+                                        struct uvwasi_fd_table_t* table,
+                                        const uvwasi_fd_t dst,
+                                        const uvwasi_fd_t src) {
+  struct uvwasi_fd_wrap_t* dst_entry;
+  struct uvwasi_fd_wrap_t* src_entry;
+  uv_fs_t req;
+  uvwasi_errno_t err;
+  int r;
+
+  if (uvwasi == NULL || table == NULL)
+    return UVWASI_EINVAL;
+
+  if (dst == src)
+    return UVWASI_ESUCCESS;
+
+  uv_rwlock_wrlock(&table->rwlock);
+
+  if (dst >= table->size || src >= table->size) {
+    err = UVWASI_EBADF;
+    goto exit;
+  }
+
+  dst_entry = table->fds[dst];
+  src_entry = table->fds[src];
+
+  if (dst_entry == NULL || dst_entry->id != dst ||
+      src_entry == NULL || src_entry->id != src) {
+    err = UVWASI_EBADF;
+    goto exit;
+  }
+
+  uv_mutex_lock(&dst_entry->mutex);
+  uv_mutex_lock(&src_entry->mutex);
+
+  /* Close the existing destination descriptor. */
+  r = uv_fs_close(NULL, &req, dst_entry->fd, NULL);
+  uv_fs_req_cleanup(&req);
+  if (r != 0) {
+    uv_mutex_unlock(&dst_entry->mutex);
+    uv_mutex_unlock(&src_entry->mutex);
```

Comment:
These unlocks and the ones at lines 383-384 look like a potential deadlock. You should normally release mutexes in reverse order of acquiring them.

Even when there's no deadlock it's probably still bad for performance because the acquirer can get rescheduled twice in case of contention.